lib/mayaUsd/nodes/proxyShapeBase.cpp

```diff
@@ -498,7 +498,7 @@ MayaUsdProxyShapeBase::computeInStageDataCached(MDataBlock& dataBlock)
         }
         else if (!usdStage) {
             // Create a new stage in memory with an anonymous root layer.
-            UsdStageCacheContext ctx(UsdMayaStageCache::Get());
+            UsdStageCacheContext ctx(UsdMayaStageCache::Get(loadSet == UsdStage::InitialLoadSet::LoadAll));
             usdStage = UsdStage::CreateInMemory("", loadSet);
         }
```

Comment:
I think you are right on both ideas. We don't need to check the value of usdStage in the else if, it should just be an else. I traced back through the code in CreateInMemory and I think the intention is we won't get another anonymous layer with  the same name so there is no point it checking the cache for it, or for putting it into the cache. I'll make those changes.

Reply:
@kxl-adsk Do you remember why we added we were using the stage cache here?

Reply:
By @kxl-adsk in another comment thread:

> It is true that currently we may not have any workflow which would require this stage to be inserted into the cache. At the same time, I don't see what could go wrong if we do insert it and keep the code symmetric with the case where we open the stage from file.

I'm not sure having the anonymous stage in the cache is a good idea. Is it possible for two different proxyShapes to open the same anonymous stage at the same time to benefit from the cache? Normally the point of the cache is to prevent loading the stage twice, but if the stage can't be shared then it doesn't help us, but maybe it can hurt us.

@seando-adsk What is the process for an anonymous stage getting a name? To avoid "leaking" stages in the cache we'd need additional code when we rename anonymous stages to remove the anonymous stage from the cache and re-add it with the new name. The cache is an associative container with the stage name as one of the possible keys, so to prevent it from becoming invalid we need the remove & re-add. Actually any time we attempt to rename a stage we'd need to remove & add, depending on how the code works.

Reply:
I had purposely not given this anonymous stage a name because our design team was still trying to sort how how to name them (taking the new Layer Editor into account). They have now given me the naming as "anonymousLayer1". So I will be making this change soon.

Reply:
If anonymousLayer1 already exists in a different proxyShape then the stage would be named anonymousLayer2 instead? If it re-uses the same name & uses the cache then whatever is already in anonymousLayer1 would show up here as well, which I don't think is intended.

@seando-adsk is there code to rename stages somewhere?

I keep on having more and more questions. What if two proxyShapes share a stage and then one of them renames it? Do they both now share the stage with the updated name? Do they have different stages? And what should be in the stage cache. This issue feels too big for my loadPayloads change here.

I think what we should do is have a separate discussion in the future when we start working on multi-proxy workflows and view this as just an incremental improvement rather than fixing all the issues with stage caching.

Reply:
I agree with your suggestion about discussing multi-proxy workflows.  Sharing stages between multiple proxy shapes feels like a form of instancing, and I know that currently there is UFE code that depends on a strict one to one mapping between proxy shapes and stages, i.e. no sharing of stages between proxy shapes.  We would need to confirm whether rendering code makes this assumption as well.

Reply:
The way we are currently using stages with in-memory root layers, I don't think we would run into a problem. BUT I see your point @williamkrick ! Let's proceed with your current change where you removed the stage cache context from the in-memory stage.

Reply:
One automated test failure lead me to the reason we _have_ to have the stage cache: the proxyShape only holds a weak reference to the `UsdStage` and so without the cache holding a strong reference the stage is deleted when the local strong pointer goes out of scope at the end of this function.

I logged https://github.com/Autodesk/maya-usd/issues/528 for the issue of the proxyShape not actually keeping its UsdStage alive, and brought back the stage cache for now.